tap_gitlab/client.py

```diff
@@ -81,32 +103,48 @@ def get_next_page_token(
         """Return token for identifying next page or None if not applicable."""
         return response.headers.get("X-Next-Page", None)
 
+    @staticmethod
+    def _url_encode(val: Union[str, datetime, bool, int, List[str]]) -> str:
+        """Encode the val argument as url-compatible string."""
+        return urllib.parse.quote_plus(str(val))
 
-class ProjectBasedStream(GitLabStream):
-    """Base class for streams that are keys based on project ID."""
+    def get_url(self, context: Optional[dict]) -> str:
+        """Get stream entity URL."""
+        url = "".join([self.url_base, self.path or ""])
+        vals = copy.copy(dict(self.config))
+        vals.update(context or {})
+        for key, val in vals.items():
+            search_text = "".join(["{", key, "}"])
+            if search_text in url:
+                url = url.replace(search_text, self._url_encode(val))
+                if "{project_path}" in search_text:
+                    self.logger.info(
+                        f"DEBUG: Found project arg. URL is {url} after parsing "
+                        f"input val '{val}' to '{self._url_encode(val)}'."
+                    )
 
-    state_partitioning_keys = ["project_path"]
+        return url
 
-    @property
-    def partitions(self) -> List[dict]:
-        """Return a list of partition key dicts (if applicable), otherwise None."""
-        if "{project_path}" in self.path:
-            if "projects" not in self.config:
-                raise ValueError(
-                    f"Missing `projects` setting which is required for the "
-                    f"'{self.name}' stream."
-                )
+    def post_process(self, row: dict, context: Optional[dict] = None) -> Optional[dict]:
+        """Post process records."""
+        result = super().post_process(row, context)
+        del row
+        if result is None:
+            return None
 
-            return [
-                {"project_path": id}
-                for id in cast(list, self.config["projects"].split(" "))
-            ]
+        assert context is not None  # All streams have parent or partitions
```

Comment:
@edgarrmondragon - Thanks for raising. I actually created these asserts for my own debugging purposes (since that condition should never be reached if basic assumptions are correct. I'll just remove those asserts, I think, since there's not a need to keep the guard/exception handling post-development.

Reply:
@edgarrmondragon - My memory was rusty on this, but after going back to the code to remove the asserts, I remembered that those asserts were also needed to inform the mypy and the linter that context is non-null. I've added a code comment to this affect - which keeps the linter happy while also being a little more clear about my reasons for using `assert` instead of try/raise.

What do you think of this approach? I've read proponent cases of `assert` being sometimes preferred for lint and interpreter hints (versus a full "try/except/raise" everywhere) but we don't have team dev standards on this as of yet.

Also - I double checked and the linter will not let me change the signature to have a non-null dict, since the parent class in the SDK expects `Optional[dict]`.

Reply:
@aaronsteers I see. I don't have a strong opinion then, but would something like this work

```python
context = context or {}
```

Reply:
I'm late to the party, but wanted to chime in on the use of `assert`. I've become a bit of a fan for a few reasons, on top of making `mypy` happy:
- `assert context is not None, "Some useful error message"` will do wonders during development to guide users of libs such as the sdk.
- those error messages can also help catch errors when first deploying (particularly handy for data-related work, where it's hard to foresee all possible inputs)
- once you're comfortable that the code is solid, running with `python -O ...` will disable `assert`s and make them no-ops, so there is virtually no performance impact.